src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -115,27 +123,46 @@ public Object execute(
     if (METHOD_SET_READ_ONLY.equals(methodName) && args != null && args.length > 0) {
       switchConnectionIfRequired((Boolean) args[0]);
     }
+
     try {
-      return this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
+      final Object result = this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
+
+      if (this.explicitlyReadOnly && this.loadBalanceReadOnlyTraffic && isTransactionBoundary(methodName)) {
+        this.inTransaction = false;
+        pickNewReaderConnection();
+      }
+      return result;
     } catch (SQLException e) {
       if (isCommunicationException(e)) {
         final JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
         final HostInfo currentHost = this.currentConnectionProvider.getCurrentHostInfo();
-        closeInternalConnection(this.readerConnection, currentConnection);
-        closeInternalConnection(this.writerConnection, currentConnection);
+        closeAllConnections(currentConnection);
         updateTopology(currentConnection, currentHost);
-        updateInternalConnections(currentConnection, currentHost);
+        updateInternalConnectionInfo(currentConnection, currentHost);
       }
       throw e;
     }
   }
 
+  private boolean isTransactionBoundary(String methodName) throws SQLException {
+    JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+    if (currentConnection == null) {
+      return false;
+    }
+
+    final boolean isAutoCommit = currentConnection.getAutoCommit();
+    if (isAutoCommit && ("execute".equals(methodName) || "executeQuery".equals(methodName))) {
```

Comment:
I'm not sure about a purpose of the following condition
("execute".equals(methodName) || "executeQuery".equals(methodName))

Reply:
If autocommit is on, we would want to switch to a new reader after every query that is executed but not necessarily on every method called on the connection. For example if getAutoCommit was called on the connection we would not want to switch. Let me know if you think I'm missing something